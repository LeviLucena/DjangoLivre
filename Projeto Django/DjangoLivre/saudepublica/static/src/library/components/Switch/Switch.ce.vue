<template>
  <div
    v-tooltip="{ text: tooltipText, place: tooltipPlace, type: tooltipType, timer: tooltipTimer }"
    class="br-switch"
    :class="classList"
  >
    <input
      v-bind="$attrs"
      :id="_id"
      type="checkbox"
      :name="name"
      :value="_value"
      :checked="_checked"
      :disabled="disabled"
      :indeterminate="isIndeterminate"
      @change="runChangeHandlers"
      @innerChange="runInnerChangeHandlers"
    />
    <label :for="_id">
      <slot>{{ label }}</slot>
    </label>
    <div
      v-if="labelChecked && labelNotChecked"
      class="switch-data"
      :data-enabled="labelChecked"
      :data-disabled="labelNotChecked"
    ></div>
  </div>
</template>

<script>
import Checkgroup from "../../mixins/checkgroup"
import TooltipMixin from "../../mixins/TooltipMixin"

const Switch = {
  name: "brSwitch",
  inheritAttrs: false,
  emits: [
    "onChange", // Evento emitido quando o estado do switch é modificado.
    "update:checked", // Evento emitido para fazer o two-way data binding com a prop checked.
    /**
     * Referência sobre o consumo de eventos:
     * [Eventos Emitidos pelos Componentes](/?path=/story/eventos-emitidos-pelos-componentes--page)
     */
  ],
  mixins: [Checkgroup, TooltipMixin],
  props: {
    /** **[OPCIONAL]** Se estiver presente, indica que o estado inicial do switch é true. */
    checked: {
      type: Boolean,
      default: false,
    },
    /** **[OPCIONAL]** Se estiver presente, indica que o switch está desabilitado. */
    disabled: {
      type: Boolean,
      default: false,
    },
    /** **[OPCIONAL]** Se estiver presente, coloca um ícone na chave de seleção para indicar a mudança de estado. */
    icon: {
      type: Boolean,
      default: false,
    },
    /** **[OPCIONAL]** Id do componente. Usado, também para conectar label correspondente. */
    id: {
      type: String,
      default: null,
    },
    /** **[OPCINAL]** Indica o tamanho do switch. */
    size: {
      type: String,
      validator: function (value) {
        return [null, "large", "small"].includes(value)
      },
      default: null,
    },
    /** **[OPCIONAL]** Descreve o que o switch faz quando a alternância estiver ativada. Se estiver presente sobrescre a label passada por slot. */
    label: {
      type: String,
      default: null,
    },
    /** **[OPCIONAL]** Rótulo para a chave ativada, que complementa a label, auxiliando no entendimento da posição em que a chave se encontra */
    labelChecked: {
      type: String,
      default: null,
    },
    /** **[OPCIONAL]** Rótulo para a chave desativada, que complementa a label, auxiliando no entendimento da posição em que a chave se encontra */
    labelNotChecked: {
      type: String,
      default: null,
    },
    /** **[OPCIONAL]** Define o atributo name que será atribuido ao switch. */
    name: {
      type: String,
      default: null,
    },
    /** **[OPCIONAL]** Label localizado acima do switch */
    top: {
      type: Boolean,
      default: false,
    },
    /** **[OPCIONAL]** Label localizado a direita do switch */
    right: {
      type: Boolean,
      default: false,
    },
  },
  data() {
    return {
      isChecked: this.checked, // [DATA] Estado atual do switch.
    }
  },
  computed: {
    /**
     * [PROPRIEDADE COMPUTADA] - Lista de classes.
     * @return {Object} Objeto contendo classes aplicadas na raiz do componente.
     */
    classList() {
      return {
        icon: this.icon,
        top: this.top,
        right: this.right,
        large: this.size === "large",
        small: this.size === "small",
      }
    },
    /**
     * [PROPRIEDADE COMPUTADA] Controla o two-way data binding do estado checked com propriedades get and set.
     * @return {Object} Objeto contendo métodos get e set.
     */
    _checked: {
      /**
       * Pega o estado checked passado ao componente. Só pode ser alterado pelo "lado de fora do switch" usando a prop "checked".
       * @return {Boolean} Prop checked.
       */
      get() {
        return this.checked
      },
      /**
       * Guarda o estado do switch no data "isChecked".
       * Emite o evento "update:checked" usado para two-way data binding.
       * @param {Boolean} value Estado checked do switch.
       */
      set(value) {
        this.isChecked = value
        this.$emit("update:checked", value)
      },
    },
    /**
     * [PROPRIEDADE COMPUTADA] Computa qual id será usado no switch.
     * @return {String} Valor da prop id se houver, senão valor computado usando o uid.
     */
    _id() {
      return this.id || "br-switch-input-" + this._.uid
    },
    /**
     * [PROPRIEDADE COMPUTADA] Computa o valor usado no atributo value do switch.
     * @return {String, Boolean} Valor do estado do switch ou das props "labelChecked" e "labelNotChecked".
     */
    _value() {
      return this.isChecked ? this.labelChecked || this.isChecked : this.labelNotChecked || this.isChecked
    },
  },
  methods: {
    /**
     * Emite evento customizado com o nome "onChange"
     */
    emitOnChangeEvent() {
      this.$emit("onChange", this.$el.querySelector("input"))
    },
    /**
     * Emite evento customizado com o nome "update:checked", usado para two-way data bindind.
     * @param {Object} event Objeto Event.
     */
    emitUpdateCheckedEvent(event) {
      this.$emit("update:checked", event.target.checked)
    },
    /**
     * Executa os handlers para o evento "change".
     * @param {Object} event Objeto Event
     */
    runChangeHandlers(event) {
      this.updateCheckedState(event)
      this.updateIndeterminateState(event)
      this.emitOnChangeEvent()
    },
    /**
     * Executa os handlers para o evento customizado "innerChange"
     * @param {Object} event Objeto Event
     */
    runInnerChangeHandlers(event) {
      this.updateCheckedState(event)
      this.updateIndeterminateState(event)
    },
    /**
     * Atualiza o estado checked do switch por meio da propriedade computada _checked.
     * @param {Object} event Objeto Event
     */
    updateCheckedState(event) {
      this._checked = event.target.checked
    },
  },
}

export default Switch
</script>

<style lang="scss">
@import "~@govbr-ds/core/src/partial/scss/_base";
@import "~@govbr-ds/core/src/partial/scss/_utilities";
@import "~@govbr-ds/core/src/components/switch/_switch";
</style>
