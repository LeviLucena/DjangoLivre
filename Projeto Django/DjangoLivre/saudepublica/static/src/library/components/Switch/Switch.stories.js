import { action } from '@storybook/addon-actions'
import { customEventStorybookArgType, kebabiseArgs } from '../../util/Utils.js'
import BrSwitch from './Switch.ce.vue'

const slotDefault = `
  Label
`
export default {
  title: 'Componentes/Switch',
  component: BrSwitch,
  argTypes: {
    checked: {
      control: 'boolean',
      defaultValue: false,
    },
    disabled: {
      defaultValue: false,
    },
    icon: {
      control: 'boolean',
      defaultValue: false,
    },
    id: {
      control: 'text',
      defaultValue: '',
    },
    size: {
      control: { type: 'select', options: ['small', 'medium', 'large'] },
    },
    label: {
      control: 'text',
      defaultValue: '',
    },
    default: {
      control: 'text',
      defaultValue: '',
      table: {
        type: {
          summary: 'NodeList',
        },
        defaultValue: {
          summary: '',
        },
      },
    },
    labelChecked: {
      control: 'text',
      defaultValue: '',
    },
    labelNotChecked: {
      control: 'text',
      defaultValue: '',
    },
    top: {
      defaultValue: false,
    },
    right: {
      defaultValue: false,
    },
    change: {
      ...customEventStorybookArgType,
    },
    tooltipText: {
      control: 'text',
    },
    tooltipPlace: {
      options: ['top', 'right', 'bottom', 'left'],
      control: 'inline-radio',
      if: { arg: 'tooltipText', truthy: true },
    },
    tooltipType: {
      options: ['info', 'success', 'warning', 'error'],
      control: 'inline-radio',
      if: { arg: 'tooltipText', truthy: true },
    },
    tooltipTimer: {
      control: 'number',
      if: { arg: 'tooltipText', truthy: true },
    },
  },
}

const Template = (args) => ({
  props: Object.keys(args),
  setup() {
    return { args: kebabiseArgs(args) }
  },
  template: `<br-switch v-bind="args" @onChange="handleChange"></br-switch>`,
  methods: {
    handleChange: action('onChange'),
  },
})

const TemplateComSlot = (args) => ({
  props: Object.keys(args),
  setup() {
    return { args: kebabiseArgs(args) }
  },
  template: `<br-switch v-bind="args" @onChange="handleChange">
    ${args.default}
  </br-switch>`,
  methods: {
    handleChange: action('onChange'),
  },
})

const defaultExcluded = ['checkgroupParent', 'checkgroupChild', 'indeterminate', 'update:indeterminate']

export const Completo = TemplateComSlot.bind({})
Completo.args = {
  icon: true,
  checked: true,
  size: 'large',
  labelChecked: 'Sucesso',
  labelNotChecked: 'Erro',
  default: slotDefault,
}
Completo.parameters = {
  controls: {
    exclude: defaultExcluded,
  },
}

export const Simples = Template.bind({})
Simples.args = {
  label: 'Simples',
}
Simples.parameters = {
  controls: {
    exclude: defaultExcluded,
  },
}

export const ComTooltip = TemplateComSlot.bind({})
ComTooltip.args = {
  icon: true,
  checked: true,
  size: 'large',
  labelChecked: 'Sucesso',
  labelNotChecked: 'Erro',
  default: slotDefault,
  tooltipText: 'Lorem ipsum dolor sit amet.',
  tooltipPlace: 'bottom',
}
ComTooltip.parameters = {
  controls: {
    exclude: defaultExcluded,
  },
}
