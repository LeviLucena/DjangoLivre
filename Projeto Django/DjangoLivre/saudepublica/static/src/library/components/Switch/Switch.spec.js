import { shallowMount } from '@vue/test-utils'
import Switch from './Switch.ce.vue'

jest.spyOn(global.console, 'error').mockImplementation((s) => {
  throw s
})

describe('br-switch', () => {
  test('Testa se o uso da classe br-switch está sendo renderizado.', () => {
    const wrapper = shallowMount(Switch, {
      propsData: {
        id: '123',
      },
    })
    expect(wrapper.find('.br-switch').exists()).toBeTruthy()
  })

  test('Testa se a prop checked está sendo renderizada.', () => {
    const wrapper = shallowMount(Switch, {
      propsData: {
        id: '123',
        checked: true,
      },
    })
    expect(wrapper.find('input').element.checked).toBeTruthy()
  })

  test('Testa se a prop disabled está sendo renderizada.', () => {
    const wrapper = shallowMount(Switch, {
      propsData: {
        id: '123',
        disabled: true,
      },
    })
    expect(wrapper.find('input').element.disabled).toBeTruthy()
  })

  test('Testa se a prop icon está sendo renderizada.', () => {
    const wrapper = shallowMount(Switch, {
      propsData: {
        id: '123',
        icon: true,
      },
    })
    expect(wrapper.find('.br-switch.icon').exists()).toBeTruthy()
  })

  test('Testa se a prop size com valor small está sendo renderizada.', () => {
    const wrapper = shallowMount(Switch, {
      propsData: {
        id: '123',
        size: 'small',
      },
    })
    expect(wrapper.find('.br-switch.small').exists()).toBeTruthy()
  })

  test('Testa se a prop size com valor large está sendo renderizada.', () => {
    const wrapper = shallowMount(Switch, {
      propsData: {
        id: '123',
        size: 'large',
      },
    })
    expect(wrapper.find('.br-switch.large').exists()).toBeTruthy()
  })

  test('Testa se a prop label está sendo renderizada.', () => {
    const textoHello = 'hello'
    const wrapper = shallowMount(Switch, {
      propsData: {
        id: '123',
        label: textoHello,
      },
    })
    expect(wrapper.text()).toMatch(textoHello)
  })

  test('Testa se as props labelChecked e labelNotChecked estão sendo renderizadas.', () => {
    const labelChecked = 'ON'
    const labelNotChecked = 'OFF'
    const wrapper = shallowMount(Switch, {
      propsData: {
        id: '123',
        labelChecked: labelChecked,
        labelNotChecked: labelNotChecked,
      },
    })
    expect(wrapper.html()).toMatch('data-enabled="' + labelChecked + '"')
    expect(wrapper.html()).toMatch('data-disabled="' + labelNotChecked + '"')
    expect(wrapper.element).toMatchSnapshot()
  })

  test('Testa se o evento change foi emitido, ao alterar o estado do switch', async () => {
    const wrapper = shallowMount(Switch, {
      propsData: {
        id: '123',
      },
    })
    await wrapper.find('input').trigger('change')
    expect(wrapper.emitted().change).toBeTruthy()
  })
})
