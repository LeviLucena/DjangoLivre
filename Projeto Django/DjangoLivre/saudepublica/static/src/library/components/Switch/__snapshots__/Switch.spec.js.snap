// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`br-switch Testa se as props labelChecked e labelNotChecked est√£o sendo renderizadas. 1`] = `
<div
  class="br-switch"
>
  <input
    id="123"
    type="checkbox"
    value="OFF"
  />
  <label
    for="123"
  >
    
    
    
  </label>
  <div
    class="switch-data"
    data-disabled="OFF"
    data-enabled="ON"
  />
</div>
`;
